swig/include/python/gdal_python.i

```diff
@@ -1580,7 +1580,23 @@ def ReleaseResultSet(self, sql_lyr):
 %}
 
 %feature("pythonappend") GetRasterBand %{
-    self._add_band_ref(val)
+    self._add_child_ref(val)
+%}
+
+%feature("pythonappend") GetLayerByName %{
+    self._add_child_ref(val)
+%}
+
+%feature("pythonappend") GetLayerByIndex %{
+    self._add_child_ref(val)
+%}
+
+%feature("pythonappend") CreateLayer %{
```

Comment:
In theory, we should likely also instrument DeleteLayer() to make sure it removes the reference to avoid the array of reference to grow without control if someone would do an infinite loop of CreateLayer() + DeleteLayer(). But that might require an auxiliary data structure to map the layer index to its object. That could be for a follow-up pull request as I don't anticipate this scenario to be a nominal use case.

Reply:
Wouldn't the `WeakSet` that we are using for the container take care of this for us?

Reply:
Indeed:

```
ds = gdal.GetDriverByName("GPKG").Create("/tmp/test.gpkg", 0, 0, gdal.GDT_Unknown)
for i in range(100):
    lyr_name = f"ds_{i}"
    lyr = ds.CreateLayer(lyr_name)
    ds.DeleteLayer(lyr_name)

assert len(ds._child_references) == 1
```

Reply:
oh, nice!